internal/app.go

```diff
@@ -31,6 +31,7 @@ type CryptoHandler interface {
 
 type App struct {
 	EncryptedConfig string
+	EncryptedBackup string
```

Comment:
I think we should change `safeWrite` to return the original file contents (if present). Then  we don't have to introduce this intermediate state file.

Reply:
I like this suggestion.
However, before doing that, could you explain why do we have a separate "extract" command in fioconfig ?
If we make this change, then an extract command wouldn't be able to leverage this proper removal fuctionality.

Reply:
I thought that if something goes wild, someone can connect to a device, and run `fioconfig extract` manually to enforce a desired state.  That's why I thought keeping a backup file should help in this case.

Reply:
> However, before doing that, could you explain why do we have a separate "extract" command in fioconfig ?

config.encrypted lives in persistent storage. all the extracted secrets are in tmpfs. Upon a reboot, we need to "extract" the secrets so they are available to the system (and network won't be up or fioconfig).

So "extract" is only run on content that's been previously pulled down via a "check-in"

Reply:
@doanac based on what I understand from your comment are these two important things:
1. We do use an extract to put a device into a proper state after a reboot.
2. A secrets folder is wiped out on reboot.

Now, imagine the following situation:
1. A device has started a check-in and pulled a new `config.encrypted`.
2. In the middle of a check-in an unexpected reboot has happened.
3. As a check-in operation is not atomic - we'd have a part of files up to date, and another part out of date.
4. After a reboot we run extract, and as a secrets folder is empty - all files are treated as changed and we run all onchanged handlers.
5. All is good, but if we don't have a backup of a previous `config.encrypted` version - those files that should have been removed will not be on disk (that's OK), but their onchanged handlers will not be called (that's not OK).

I understand that our contrib handlers may be ignorant of this issue, but in general this is still a bug.
Thus, because a check-in operation isn't atomic, I'd prefer to keep a backup file on disk, so that we can properly restore a device state after an unexpected reboot.

Also, thanks to your comment I've found a small defect in my code (also related to a reboot) which I'm going to fix, as described in the other comment.

Reply:
Something still looks wrong that we need two state files. I get nervous overthinking this as handlers weren't intended to take on this much responsibility, but I think your real bug is this: We are committing config.encrypted to disk *before* we've actually set things in place. If we write this file after doing the handlers, then things work properly. In the case of a reboot, things could self-correct.

Reply:
I think that's doable.  Each solution has it's pros and cons, and is anyway not ideal, so I'll try implementing what you think is a better one.

Reply:
Done
Fortunately, existing tests has passed with almost zero changes.
There were some non-trivial changes needed to several functions, but not really special.